src/pycram/process_module.py

```diff
@@ -3,132 +3,350 @@
+    :return: The decorated function wrapped into the decorator
+    """
     def wrapper(*args, **kwargs):
-        pre = ProcessModule.robot_type
-        ProcessModule.robot_type = "real"
+        pre = ProcessModuleManager.execution_type
+        ProcessModuleManager.execution_type = "real"
         func(*args, **kwargs)
-        ProcessModule.robot_type = pre
+        ProcessModuleManager.execution_type = pre
+
     return wrapper
 
 
 def with_simulated_robot(func: Callable) -> Callable:
+    """
+    Decorator to execute designators in the decorated class on the simulated robot.
+
+    Example:
+
+    .. code-block:: python
+
+        @with_simulated_robot
+        def plan():
+            some designators
+
+    :param func: Function this decorator is annotating
+    :return: The decorated function wrapped into the decorator
+    """
     def wrapper(*args, **kwargs):
-        pre = ProcessModule.robot_type
-        ProcessModule.robot_type = "simulated"
+        pre = ProcessModuleManager.execution_type
+        ProcessModuleManager.execution_type = "simulated"
         func(*args, **kwargs)
-        ProcessModule.robot_type = pre
+        ProcessModuleManager.execution_type = pre
+
     return wrapper
 
 
-# These are imported so they dont have to be initialized when executing with
+# These are imported, so they don't have to be initialized when executing with
 simulated_robot = simulated_robot()
 real_robot = real_robot()
+
+
+class ProcessModuleManager(ABC):
+    """
+    Base class for managing process modules, any new process modules have to implement this class to register the
+    Process Modules
+    """
+    execution_type = None
+    """
+    Whether the robot for which the process module is intended for is real or a simulated one
+    """
+    available_pms = []
+    """
+    List of all available Process Module Managers
+    """
+    _instance = None
+    """
+    Singelton instance of this Process Module Manager
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Creates a new instance if :py:attr:`~ProcessModuleManager._instance` is None, otherwise the instance
+        in :py:attr:`~ProcessModuleManager._instance` is returned.
+        :return: Singelton instance of this Process Module Manager
+        """
+        if not cls._instance:
+            cls._instance = super(ProcessModuleManager, cls).__new__(cls)
+            return cls._instance
+        else:
+            return cls._instance
+
+    def __init__(self, robot_name):
+        """
+        Registers the Process modules for this robot. The name of the robot has to match the name given in the robot
+        description.
+
+        :param robot_name: Name of the robot for which these Process Modules are intended
+        """
+        self.robot_name = robot_name
+        ProcessModuleManager.available_pms.append(self)
+
+    @staticmethod
+    def get_manager() -> ProcessModuleManager | None:
```

Comment:
whats the different of this to the init/new method? 
From a users perspective it might be confusing to have different behaviors with MySingleton() and MySingleton.get_manager().
Imo having the default construction call do all the work is the best regarding usability.